[driver] Added TCS3472 missing features

TomSaw opened the pull request:
Significantly improves possible performance. Got sample-rates up to 200Hz now:

### Bug
- shrinked commandBuffer[4] -> commandBuffer[3]. Until now, the driver send a placeholder on every request witch seems not to be necessary.

### Features
- Interrupt control
   - enable and reload Interrupts
   - set interrupt persistence filter
   - set interrupt thresholds for the comparator that triggers interrupts
   - getter for clear value - (Cause the interrupt-comparator is wired to clear value, it's good to inspect actual readings to setup comparator-threesholds)
- set and enable waittime (delay between samples, conceived for energy savings)

_Sry for the reformatted code... if that's a problem, i'll fix it._

salkinium commented:
> Sry for the reformatted code... if that's a problem, i'll fix it.

1. We use tabs (unfortunately), not spaces. That would already reduce the noise.
2. There is a custom `.clang-format` config here that you can use instead of formatting manually. https://github.com/modm-io/modm/blob/develop/.clang-format Or at least you can run it once over the file you want and then change the things that you don't like (sometimes the "strict" formatting rules are not that good).

Update: There's also a `tools/scripts/clang-format.sh` for only updating the files you changed: https://github.com/modm-io/modm/blob/develop/tools/scripts/clang-format.sh
Alternatively there's a `git clang-format origin/develop` that does that same but better? https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format

TomSaw commented:
Thanks for the rich options!

I'm fine with the modm coding-style ;) so a simple IDE-plugin, applying the .clang-format from within the working-folder, did the job.

It's dropping lovely handmade vertical alignments of Register enums but i don't care. If you don't mind either, the formatting looks good now.

salkinium commented:
The `ext/dlr/scons-build-tools` submodule needs to be removed in your branch, you want to remove the folder and alaos entry from `.gitmodules`.

salkinium commented:
Ok, so my understanding from googling around is that TCS3414 has been discontinued, and been replaced by TCS3472.

- [Datasheet TCS3414](https://media.digikey.com/pdf/Data%20Sheets/Austriamicrosystems%20PDFs/TCS3404_14_Rev_Nov_2015.pdf)
- [Datasheet TCS3472](https://ams.com/documents/20143/36005/TCS3472_DS000390_3-00.pdf/6fe47e15-e32f-7fa7-03cb-22935da44b26)

But these are not compatible devices, ie. I cannot use the TCS3472 driver with a TCS3414 device, for example the register map is different, and also the register contents are different:
<details>
<summary>TCS3414 register map</summary>

![](https://user-images.githubusercontent.com/163066/110943682-cc5eb500-833b-11eb-8181-561f146dcc63.png)

</details>
<details>
<summary>TCS3472 register map</summary>

![](https://user-images.githubusercontent.com/163066/110943707-cf59a580-833b-11eb-8f62-f478f822baf2.png)

</details>

So I would recommend moving your TCS3472 driver into a new lbuild module `:driver:tcs3472` (copy paste) and leave the TCS3414 as is.

salkinium commented:
Oh, wait, I'm an idiot. You updated *both* drivers. Never mind me. 🤦‍♂️🤦‍♂️

TomSaw commented:
Not exactly.. i did not update the tcs3414... but if i compare the branches i see, i've fragged tcs3414.hpp 😅
Update: tcs3414.hpp is restored from upstream/develop

TomSaw commented:
Oh and excuse this additional DLR submodule... next time i'll take better care

salkinium commented:
Ok, since I had this sensor lying around and always wanted to test it, I added an example and tested it in hardware.

I also refactored the driver some more to make it more up-to-date to our coding style and added the Data class mechanism that we typically use now. It separates the data storage and representation from the driver, which makes it easy to copy around and transmit over a network and only does the conversion work (like `getNormalizedColor()`) when it's necessary. I'll do the same with the TCS3414 driver tomorrow.

Could you test my changes in your hardware when you have time?

TomSaw commented:
Good morning! I'm currently reading your changes and will test on my hardware soon.

TomSaw commented:
Tested and works. The new 'Data mechanism' looks nice and clean.
Thanks for the adjustments, i will update my other code too.

TomSaw committed:
[driver] Refactor TCS34xx color sensors

- Use separate Data class mechanism.
- Add threshold, interrupt and wait time setters to TCS3472

Co-authored-by: Niklas Hauser <niklas.hauser@rwth-aachen.de>

TomSaw committed:
[examples] Add TCS3472 color sensor examples

salkinium reviewed (approved):
I'll squash the commits together when you tell me you're ready (and the CI passes).

TomSaw commented:
I'm ready, ... but
how about lining up another example for Atmega328p / Arduino-Nano (probably without threads).
I've developed on a A-Nano so only some cleaning and decoration is required.

 let me know if you think it's worth the time.

salkinium commented:
Yes, go for it. You can keep the protothreads, they cost next to nothing even on AVRs.

salkinium commented:
I'm sorry, did you mean that *I* or that *you* will add another AVR example?

TomSaw commented:
This Sensor is very popular for beginners. They f.e. build [candy-sorting machines](https://www.youtube.com/watch?v=yAMSrrVGrco) of it (they work soooooooo 🐌 slow).
My thought was: **that's the arduino-nano example, man** let's not scare the folks and keep (at least) one example **simple + blocking + polling** and take the chance to give a lesson in performance:

I want to ad a second example using tcs3472s interrupt-feature, explaining: This alone frees up tons of cpu-resources.
The third example introduces protothreads, explaining the advantage of non-blocking i2c-transactions.

But maybe, that's over-ambitios 😓: no arduino-minded person will ever stumble over the lines.

Okay, i'll rest in peace with example 3 ... coming soon.

salkinium commented:
I also think having one example with interrupts and protothreads is best.

I don't want to demotivate you, however, our examples are regrettably not very good tutorials, often they are under-documented and at the moment we mostly (ab)use them as manual integration tests. It's something I would like to change, however, my ToDo-List is quite long and there are always more important things on it… 🙃

I would perhaps recommend doing a tutorial like this as a blog post on your homepage, since that absolves you from the "pressure" of contributing to modm and you can set the level of detail and the structure of the tutorial completely by yourself. We could then link to this in the protothread module docs.

TomSaw commented:
Good recommendation. Await 2 examples later today...

TomSaw commented:
So here's one example using threads and interrupt. I'm 99% sure this works but can't test now cause my one and only Nano just got fail-fused 💀 -> reanimation failed. This is defacto my facepalm-emoji-moment...  but i'm to lazy to look for it 😁

Fresh Nanos have been ordered for testing.

TomSaw commented:
Fresh head 🧠 and got my Arduino-Nano board back to life.

- The rich example is tested and done ✅

Go, merge the stuff into develop 😁. Thanks for your guidance, and excuse this chaotic (noizy) contribution-trip. I've learned a lot and looking forward to my request.

salkinium reviewed (approved):
Squashed into the previous commit and removed the `while(!` thing for clarity.

salkinium commented:
Apologies for the CI, it usually takes 20-30mins, not 5h. CircleCI seems to have significant issues, we're in contact with their support.

salkinium commented:
> Thanks for your guidance, and excuse this chaotic (noizy) contribution-trip. I've learned a lot and looking forward to my request.

You're welcome! I'm always glad to help new contributors navigate these complex waters. There's a lot of bullshit gatekeeping in embedded, I'm not a fan of it…

TomSaw commented:
True, it requires various experience-loops to finally keep repo-things horizontal. But i appreciate the requirement and helpfulness of most of the stumbling blocks.

Manually checking the result by wrapping _while(! ... )_ is not required 😵 the better it is 😏. I'll study the protothreads again, with new eyes.

Damn, i'm glad having contributed a little - but in my scale remarkable - piece to my favorite framework 😇
